_drafts/hw_actors.md

```diff
@@ -0,0 +1,150 @@
+- send messages to other actors
+- read/write its local state
+
+Despite the actor and its mailbox are tightly coupled in this model, I think it is less practical than relying on separate queues. The latter case is more flexible since an actor may choose which queue to poll depending on the message that is being processed.
+
+Since there are no threads, the main() purpose is just an initialization. It ends with an infinite loop waiting for interrupts. Note that after initialization completes, there are no other  executable entities except interrupt handlers.
+
+Interrupts are split into two classes: interrupts producing messages and interrupts consuming messages. Also, there are only three types of software objects: queues, actors and messages.
+
+It is expected that interrupts designated to peripheral devices will post messages to queues which in turn have subscribed actors, so, pseudocode for the interrupt handler may be written as:
+
+
+        device_interrupt_handler() {
+            message = alloc_message(pool)
+            ... read device registers and set message payload ...
+            push_to_queue(queue, message)
+        }
+
+
+Interrupt handlers that are dedicated to actors contain just a call to function, examining a list of active actors with priority corresponding to that vector:
+
+
+        actor_interrupt_handler() {
+            schedule_actors(VECTOR_ID)
+        }
+
+
+On initialization phase, each actor subscribes to some queue. Once message arrives to that queue the actor is activated. Since each actor may be subscribed to exactly one queue, subscription may be implemented as return value: the actor should return a valid pointer to the queue it wants to be subscribed to. Return values may change during an actor’s lifetime but must always be a valid queue. This is because of the absence of internal queue: when an actor has not been subscribed to a queue, it will have no chance to run again. So, pseudocode for an actor is shown below:
+
+
+        actor(self, message) {
+            ... message handling ...
+            free(message)
+            return queue_to_be_subscribed_to
+        }
+
+
+Queues may contain both messages and actors. If someone posts a message in an empty queue, it will save that message until some actor will try to poll. On the other hand, if an actor tries to get a message from an empty queue it will be saved as a subscriber and later when a message arrives the actor will be activated. Therefore, a queue at any moment of time may contain either messages or actors, but never both. Note that queue is just a head of linked list, it has no internal buffers, so queue overflow cannot happen.
+
+For simplicity, messages are fixed-sized. Message allocator is just a pool of fixed-sized memory chunks. For latency-critical systems, it may be even implemented as lock-free. When the message pointer arrives to an actor, it is assumed that message type is known to its recipient since it knows what queue it polls.
```

Comment:
Would it make more sense for this to just be a fixed size queue? Then we could avoid allocations. Or is that what you're getting at?

Reply:
When allocation happens inside the queue implementation it would require copying user-provided message payload into the internal storage. While it may be OK I mostly focus on speed and simplicity in this article.